### PR TITLE
Add FAQ to README.md

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -173,11 +173,6 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
 git machete github create-pr [--draft]
 ```
 
-Synchronize with the remote repository: checkout your open PRs, delete unmanaged branches and also delete untracked managed branches with no downstream branch: <br/>
```
Comment (PawelLipski) [inline README.md:176]: Sorry... this subcommand is too niche to include in README 😅

File: README.md [hunk 2]
```diff
@@ -187,18 +182,76 @@ This will be resolved from the first of:
 
 <br/>
 
+
+## FAQ
+
+#### I've run `git machete discover`... but the branch layout I see in `.git/machete` doesn't exactly match what I expected. Am I doing something wrong?
```
Comment (PawelLipski) [inline README.md:188]: h4 and not h3 here... the latter looked weird for such long header text

File: README.md [hunk 3]
```diff
@@ -10,8 +10,8 @@
 [![CircleCI](https://circleci.com/gh/VirtusLab/git-machete/tree/master.svg?style=shield)](https://app.circleci.com/pipelines/github/VirtusLab/git-machete?branch=master)
 [![codecov](https://codecov.io/gh/VirtusLab/git-machete/branch/master/graph/badge.svg)](https://codecov.io/gh/VirtusLab/git-machete)
 
+[//]: # (The image is referenced by its full URL to ensure it renders correctly on https://pypi.org/project/git-machete/)
```
Comment (PawelLipski) [inline README.md:13]: Just so that IntelliJ shows it as a comment ;p

File: README.md [hunk 4]
```diff
@@ -10,8 +10,8 @@
...
 <img src="https://raw.githubusercontent.com/VirtusLab/git-machete/master/graphics/logo_with_name.svg" style="width: 100%; display: block; margin-bottom: 10pt;" />
-<!-- The image is referenced by its full URL to ensure it renders correctly on https://pypi.org/project/git-machete/ -->
```
Comment (PawelLipski) [inline README.md:14]: Support for HTML comments probably isn't a part of standard Markdown? just a GitHub extension? dunno

File: README.md [hunk 5]
```diff
@@ -187,18 +182,77 @@ This will be resolved from the first of:
 
 <br/>
 
+
+## FAQ
+
+#### I've run `git machete discover`... but the branch layout I see in `.git/machete` doesn't exactly match what I expected. Am I doing something wrong?
+
+[//]: # (For how to find Medium header anchors, see https://www.freecodecamp.org/news/how-to-link-to-a-specific-paragraph-in-your-medium-article-2018-table-of-contents-method-e66595fea549/)
```
Comment (PawelLipski) [inline README.md:190]: Every time I do anything with Medium, I _discover_ (no pun intended) that it sucks even more than I thought last time.

Yes, Medium does NOT have regular `<a>` named anchors in headers:

image

At least there's some way around that... copy the id from `h1` 😅

File: README.md [hunk 6]
```diff
@@ -187,18 +182,77 @@ This will be resolved from the first of:
...
+No! It's all right, `discover` is based on an (imperfect)
+[heuristic](https://medium.com/virtuslab/git-machete-strikes-again-traverse-the-git-rebase-jungle-even-faster-with-v2-0-f43ebaf8abb0#0544)
```
Comment (PawelLipski) [inline README.md:192]: Although, to be fair, Medium allows also linking to individual paragraphs (`<p>`) via these ids, like here... so maybe not so bad this time!

Comment (kgadek) [reply README.md:192]: I have a feeling that those anchors could change without warning in the future…

File: README.md (hunk 1 above)
Comment (kgadek) [reply README.md:176]: Wow, I didn't notice this command. Nice!

Yet I do agree that it's bit too niche.

File: README.md [hunk 7]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+which usually yields branch layout close to what the user expected...
+but it still might differ in details (like, branches discovered to be children of `main`/`develop` instead of each other).
```
Comment (kgadek) [inline README.md]: Overall I like the casual style of FAQ. However, I'd slightly reword this to avoid unnecessary parenthesis. Maybe even split into shorter sentences.

Maybe sth like this:
```suggestion
No! It's all right, `discover` is based on an (imperfect)
[heuristic](https://medium.com/virtuslab/git-machete-strikes-again-traverse-the-git-rebase-jungle-even-faster-with-v2-0-f43ebaf8abb0#0544)
which usually yields branch layout close to what the user would expect.
It still might not be perfect and — for example — declare branches to be children of `main`/`develop` instead of themselves.
```

File: README.md [hunk 8]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+
+Just run [`git machete edit`](https://git-machete.readthedocs.io/en/stable/#edit) to fix the layout manually.
+If you're working on JetBrains IDEs, you can use [git-machete IntelliJ plugin](https://github.com/VirtusLab/git-machete-intellij-plugin#git-machete-intellij-plugin)
+for features like branch name completion when editing `.git/machete` file.
```
Comment (kgadek) [inline README.md]:
```suggestion
to get branch name completion.
```

File: README.md [hunk 9]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+
+Also, consider [`git machete github checkout-prs`](#github-integration) instead of `git machete discover` if you already have GitHub PRs opened.
+
+<br/>
+
+#### Can I use `git merge` for dealing with stacked PRs?
+
+Generally, there's a [trilemma](https://en.wikipedia.org/wiki/Trilemma) (_choose at most two out of three_):
```
Comment (kgadek) [inline README.md]: I'm a fan of fancy words, but not always in documentation for general public. Sometimes it's fine, for example:

> Rust is a systems language pursuing the trifecta: safety, concurrency, and speed.
> — https://this-week-in-rust.org/blog/2017/10/24/this-week-in-rust-205/

Even though it is imho fine, sounds cool and is rather understandable, they've ditched it for simpler:

> Rust is a programming language empowering everyone to build reliable and efficient software.
> — https://this-week-in-rust.org/blog/2022/07/20/this-week-in-rust-452/

```suggestion
Generally speaking, a Git management system can reliably provide two out of three flows:
```

File: README.md [hunk 10]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+1. stacked PRs
+2. rebase/squash (generally: rewriting git history)
```
Comment (kgadek) [inline README.md]: I don't understand what `stacked PRs` mean in this context.

File: README.md [hunk 11]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+You might end up with a very tangled history very quickly, and a non-trivial sequence of `git cherry-pick`s might be needed to restore order.
+
+We recommend choosing only (1.) and (2.) from the trilemma, i.e. stacking PRs and using squash/rebase,
+and **not** using `git merge` (other than for [backporting hotfixes](https://slides.com/plipski/git-machete/#/11)).
+
+<br/>
+
+#### Sometimes when I run `update` or `traverse`, too many commits are taken into the rebase... how to fix that?
+
+Contrary to the popular misconception, git doesn't have a notion of ["commits belonging to a branch"](https://slides.com/plipski/git-internals#/41).
```
Comment (kgadek) [inline README.md]: Is it indeed a _popular_ misconception?

File: README.md [hunk 12]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+
+While git-machete supports merging parent branch (like `main`) to update a branch
+([`git machete traverse --merge`](https://git-machete.readthedocs.io/en/stable/#traverse)),
+it works poorly with stacked PRs.
+You might end up with a very tangled history very quickly, and a non-trivial sequence of `git cherry-pick`s might be needed to restore order.
+
+We recommend choosing only (1.) and (2.) from the trilemma, i.e. stacking PRs and using squash/rebase,
+and **not** using `git merge` (other than for [backporting hotfixes](https://slides.com/plipski/git-machete/#/11)).
+
+<br/>
```
Comment (kgadek) [inline README.md:218]: I think I'd slightly restructure this:

- explain what are the 3 choices
- "2 out of 3"
- discuss what works & what doesn't work
- describe the recommended flow

File: README.md [hunk 13]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+
+#### Sometimes when I run `update` or `traverse`, too many commits are taken into the rebase... how to fix that?
+
+Contrary to the popular misconception, git doesn't have a notion of ["commits belonging to a branch"](https://slides.com/plipski/git-internals#/41).
+A branch is just a movable pointer to a commit.
```
Comment (kgadek) [inline README.md]: This is perfectly understandable by somebody who understands C & Git already, while oblique for anybody else.
I'm not sure if this is the right place to describe what a Git-branch is and I think these paragraphs work better without this sentence.

```suggestion
```

File: README.md (hunk 11 above)
Comment (kgadek) [inline README.md]: I'd link to the amazing [Pro Git book](https://git-scm.com/book/en/v2) instead

File: README.md [hunk 14]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+This makes it hard in general case to determine the range of commits that form the "unique history" of the given branch.
+There's an entire algorithm in git-machete for determining the
+[_fork point_](https://medium.com/virtuslab/make-your-way-through-the-git-rebase-jungle-with-git-machete-e2ed4dbacd02#1ac9)
+of the branch (i.e. the place after which the unique history of the branch starts).
```
Comment (kgadek) [inline README.md:229]: Too many details imho. Simpler & shorter would imho work better:

```suggestion
This makes it hard to figure-out in all of the cases.
```

File: README.md [hunk 15]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+
+One thing that you can do to help fork-point algorithm in its job, is to **not** delete local branches instantly after they're merged or discarded.
```
Comment (kgadek) [inline README.md]:
```suggestion
One thing that you can do to help fork-point algorithm in its job, is to keep local branches for some time after they were merged or discarded.
```

File: README.md [hunk 16]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+They (or specifically, their [reflogs](https://virtuslab.github.io/tips/#git/git-reflog)) will be still useful for a while
+to determine fork points for other branches (and thus, the range of commits taken into rebase).
+
+Also, you can always override fork point for a branch explicitly
+with [`git machete fork-point --override-to...`](https://git-machete.readthedocs.io/#fork-point) flags.
```
Comment (kgadek) [inline README.md]:
```suggestion
with [`git machete fork-point --override-to...`](https://git-machete.readthedocs.io/#fork-point) command.
```

File: git_machete/docs.py [hunk 17]
```diff
@@ -302,7 +302,8 @@
         Note that the above editor selection only applies for editing the definition file,
         but not for any other actions that may be indirectly triggered by git-machete, including editing of rebase TODO list, commit messages etc.
 
-        The definition file can be always accessed and edited directly under the path returned by `git machete file` (currently fixed to <git-directory>/machete).
+        The definition file can be always accessed and edited directly under the path returned by `git machete file`
+        (which is `.git/machete`, unless worktrees or submodules are involved).
```
Comment (kgadek) [inline git_machete/docs.py]:
```suggestion
        (usually `.git/machete`, unless worktrees or submodules are involved).
```

File: docs/source/cli_help/edit.rst [hunk 18]
```diff
@@ -26,7 +26,8 @@ and selecting the first one that is defined and points to an executable file acc
 Note that the above editor selection only applies for editing the definition file,
 but not for any other actions that may be indirectly triggered by git machete, including editing of rebase TODO list, commit messages etc.
 
-The definition file can be always accessed and edited directly under the path returned by ``git machete file`` (currently fixed to <git-directory>/machete).
+The definition file can be always accessed and edited directly under the path returned by ``git machete file``
+(which is ``.git/machete``, unless worktrees or submodules are involved).
```
Comment (kgadek) [inline docs/source/cli_help/edit.rst]:
```suggestion
(usually ``.git/machete``, unless worktrees or submodules are involved).
```

File: README.md (hunk 8 above)
Comment (amalota) [reply README.md]:
```suggestion
to have branch name completion when editing `.git/machete` file.
```

File: README.md (hunk 9 above)
Comment (amalota) [reply README.md]: Actually i would skip 
```
Generally, there's a [trilemma](https://en.wikipedia.org/wiki/Trilemma) (choose at most two out of three):
1. stacked PRs
2. rebase/squash (generally: rewriting git history)
3. merge commits (i.e. commits [with 2+ parents](https://slides.com/plipski/git-internals#/7))
```
The 2 paragraphs below explain it very well and this is just unnecessary. (Just a small change to the second paragraph)

File: README.md [hunk 19]
```diff
@@ -173,40 +173,96 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+2. rebase/squash (generally: rewriting git history)
+3. merge commits (i.e. commits [with 2+ parents](https://slides.com/plipski/git-internals#/7))
+
+While git-machete supports merging parent branch (like `main`) to update a branch
+([`git machete traverse --merge`](https://git-machete.readthedocs.io/en/stable/#traverse)),
+it works poorly with stacked PRs.
+You might end up with a very tangled history very quickly, and a non-trivial sequence of `git cherry-pick`s might be needed to restore order.
+
+We recommend choosing only (1.) and (2.) from the trilemma, i.e. stacking PRs and using squash/rebase,
+and **not** using `git merge` (other than for [backporting hotfixes](https://slides.com/plipski/git-machete/#/11)).
```
Comment (amalota) [inline README.md]:
```suggestion
We recommend using `git squash/rebase` over `git merge` with stacked PRs, although you should still use `git merge` for [backporting hotfixes](https://slides.com/plipski/git-machete/#/11)).
```

File: README.md (hunk 15 above)
Comment (amalota) [reply README.md]: I actually like original version better, in this case **not** deleting carries stronger meaning and it's easier to remember.

File: README.md (hunk 11 above)
Comment (amalota) [inline README.md]: Hmm, i can sense that this is a very educational answer which introduces user to the git/git-machete issues and i like it but i agree with @kgadek that this might not be necessary and the shorter version will suffice.

File: README.md (hunk 1 above)
Comment (PawelLipski) [reply README.md:176]: @amalota's idea ;)

File: README.md (hunk 6 above)
Comment (PawelLipski) [reply README.md:192]: Yeah, that's why stable section identifiers generated in a predictable way out of section title (as in most markdown renderers) is what the _civilized world_ does... but not Medium ofc

File: README.md (hunk 8 above)
Comment (PawelLipski) [reply README.md]: I'd favor @amalota's version here... would you even get rid of a mention of ``.git/machete file` @kgadek ?

File: README.md (hunk 7 above)
Comment (PawelLipski) [reply README.md]: Hmm I'd stick to `each other` rather than `themselves`, as `themselves` suggests that _branch X is a child of X itself_... WDYT?

File: README.md (hunk 11 above)
Comment (PawelLipski) [reply README.md]: I believe it is (from my personal experience talking to devs)... but I have no hard data to prove its popularity 😅

File: README.md (hunk 15 above)
Comment (PawelLipski) [reply README.md]: +1 for @amalota, this is what I specifically mean by _keeping_ here: _not deleting_. _Keeping_ feels too ambiguous in this context

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -177,3 +177,5 @@ Review fixes should be pushed on separate commits for easier viewing on GitHub (
 
    If not, please apply changes on Medium to keep consistency.
    Since Medium does not offer conversion directly from Markdown, copy the formatted blog text from a GitHub and paste it into the Medium rich text editor.
+
+   Once the changes are applied, make sure that the `<p>`/`<h1>` ids referenced in links from README did not change (or are updated accordingly).
```
Comment (mkondratek) [inline CONTRIBUTING.md]: hm, can we refer to it more directly? smth like [README](README.md) ?

File: README.md (hunk 19 above)
Comment (PawelLipski) [reply README.md]: Huh there's still no `git squash` as of git 2.37... GitHub has Squash button, and also `git machete squash` exists. Lemme apply the suggestion with a slight modification

File: README.md (hunk 7 above)
Comment (mkondratek) [reply README.md]: yes

File: README.md (hunk 8 above)
Comment (kgadek) [reply README.md]: Yep, it's better

File: README.md (hunk 11 above)
Comment (kgadek) [reply README.md]: You're actually also right — this is educational. Originally, I've conceived an idea that FAQ is for q&a about use-cases, so is short and to the point.

However, I see and do agree that educational merit makes a good FAQ. Which one is suitable here? _I don't know_, either is fine.

In this light, please consider all my previous comments to be with "to the point" idea and take into account that now I'm on board with "educational" idea too!

------

Hmm, this actually makes a nice question: are both FAQ? If yes, do we mix them or have "FAQ short" and "FAQ edu" separate? Maybe one FAQ but using `<details>…` sections to expand? Or do we decide on one?

Comment (PawelLipski) [reply README.md]: Judging from my own experience when reading various FAQs... I'd rather see _FAQ edu_ rather than _FAQ short_! I usually take a look at FAQ when I already have some idea on what the tool does, and want to get some extra understanding. For instance, I remember [SLF4J FAQ](https://www.slf4j.org/faq.html) very warmly... and that's definitely an _edu_, not _short_ kind

File: README.md (hunk 14 above)
Comment (PawelLipski) [reply README.md:229]: Huh do you think this can stay as is, assuming the _edu_ variant?

File: README.md (hunk 13 above)
Comment (PawelLipski) [reply README.md]: Huh again, in the spirirt of _FAQ edu_... maybe retain the sentence, but use a _reference_ instead of C-specific _pointer_? 🤔

File: README.md (hunk 10 above)
Comment (PawelLipski) [reply README.md]: A slight example added, pls check if looks sufficient to you

File: README.md (hunk 12 above)
Comment (PawelLipski) [reply README.md:218]: Huh as this is contrary to your @amalota comment above https://github.com/VirtusLab/git-machete/pull/598#discussion_r930946070 which advocates for totally removing the mention of the 3 choices... TBH I'd favor @kgadek's option here, in the spirit of making an _FAQ edu_ and not _FAQ short_ 🤔

File: README.md (hunk 9 above)
Comment (PawelLipski) [reply README.md]: (TBD as there is an alternative suggestion below)

File: README.md (hunk 13 above)
Comment (amalota) [reply README.md]: @PawelLipski Yeah, as long as we agree that this is FAQ edu i would retain the sentence but swap `pointer` for `reference`

File: README.md (hunk 14 above)
Comment (amalota) [reply README.md:229]: @PawelLipski Yes, if it's the edu variant it can stay

File: README.md (hunk 9 above)
Comment (amalota) [reply README.md]: @PawelLipski I like the suggested change by @kgadek, it's much easier to understand.

Comment (amalota) [reply README.md]: Actually, since the question is regarding usage of `git merge` for stacked PRs, the `2 out of 3` is unnecessary since we know we deal with stacked PR's so there are only 2 choices: rebase/squash or merge.

Comment (amalota) [reply README.md]:
```suggestion
Generally speaking, a Git management system can reliably deal with stacked PRs (e.g. PR \#1: `develop <- foo` and PR \#2: `foo <- bar`) in one of the two ways:
```

File: README.md [hunk 21]
```diff
@@ -173,40 +173,95 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+Also, consider [`git machete github checkout-prs`](#github-integration) instead of `git machete discover` if you already have GitHub PRs opened.
+
+<br/>
+
+#### Can I use `git merge` for dealing with stacked PRs?
+
+Generally, there's a [trilemma](https://en.wikipedia.org/wiki/Trilemma) (_choose at most two out of three_):
+1. stacked PRs (e.g. PR \#1: `develop <- foo` and PR \#2: `foo <- bar`)
+2. rebase/squash (generally: rewriting git history)
+3. merge commits (i.e. commits [with 2+ parents](https://slides.com/plipski/git-internals#/7))
```
Comment (amalota) [inline README.md]:
```suggestion
1. rebase/squash (rewriting git history)
2. merge commits (i.e. commits [with 2+ parents](https://slides.com/plipski/git-internals#/7))
```

File: README.md [hunk 22]
```diff
@@ -173,40 +173,95 @@ Create the PR, using the upstream (parent) branch from `.git/machete` as the bas
...
+
+While git-machete supports merging parent branch (like `main`) to update a branch
+([`git machete traverse --merge`](https://git-machete.readthedocs.io/en/stable/#traverse)),
+it works poorly with stacked PRs.
+You might end up with a very tangled history very quickly, and a non-trivial sequence of `git cherry-pick`s might be needed to restore order.
+
+We recommend using `git squash/rebase` over `git merge` with stacked PRs, although you should still use `git merge` for [backporting hotfixes](https://slides.com/plipski/git-machete/#/11)).
```
Comment (amalota) [inline README.md]:
```suggestion
That is why we recommend using `git squash/rebase` over `git merge` with stacked PRs, although you should still use `git merge` for [backporting hotfixes](https://slides.com/plipski/git-machete/#/11)).
```

File: README.md (hunk 9 above)
Comment (PawelLipski) [reply README.md]: ☝🏻 lemme slightly reword that (hopefully making that even simpler)...

File: README.md (hunk 12 above)
Comment (PawelLipski) [reply README.md:218]: Okay, resolved somehow in the middle (hopefully)

File: README.md (hunk 14 above)
Comment (PawelLipski) [reply README.md:229]: Okay, leaving roughly as is then

File: README.md (hunk 21 above)
Comment (PawelLipski) [reply README.md]: Reworked in a slightly different manner
